### PR TITLE
[X86][AMX] Support AMX-TRANSPOSE

### PR DESCRIPTION
**Pull request by phoebewang**
Ref.: https://cdrdv2.intel.com/v1/dl/getContent/671368

### CODE REVIEW

File: llvm/include/llvm/CodeGen/TileShapeInfo.h [hunk 1]
```diff
@@ -34,9 +34,31 @@ class ShapeT {
     if (MRI)
       deduceImm(MRI);
   }
+  // When ShapeT has mult shapes, we only use Shapes (never use Row and Col)
+  // and ImmShapes. Due to the most case is only one shape (just simply use
+  // Shape.Row or Shape.Col), so here we don't merge Row and Col into vertor
```
[the remark this answers is not in the document]
Comment (phoebewang) [reply llvm/include/llvm/CodeGen/TileShapeInfo.h]: Done, thanks!

File: llvm/lib/Target/X86/X86ExpandPseudo.cpp [hunk 2]
```diff
@@ -568,6 +568,131 @@ bool X86ExpandPseudo::expandMI(MachineBasicBlock &MBB,
...
+        MF->getMachineMemOperand(OldMMO, TmmSize, TmmSize);
+
+    MIBLo.setMemRefs(MMOLo);
+    MIBHi.setMemRefs(MMOHi);
+
+    // Delete the pseudo.
+    MBB.erase(MBBI);
+    return true;
+  }
+  // Smilar with TILEPAIRLOAD, TILEPAIRSTORE is just for TILEPair spill, no
```
[the remark this answers is not in the document]
Comment (phoebewang) [reply llvm/lib/Target/X86/X86ExpandPseudo.cpp]: Done, thanks!

File: llvm/lib/Target/X86/X86FastTileConfig.cpp [hunk 3]
```diff
@@ -80,28 +80,41 @@ INITIALIZE_PASS_BEGIN(X86FastTileConfig, DEBUG_TYPE,
...
 
   if (MO.isReg()) {
     Register Reg = MO.getReg();
-    // FIXME it may be used after Greedy RA and the physical
+    // FIXME: It may be used after Greedy RA and the physical
     // register is not rewritten yet.
-    if (Reg.isVirtual() &&
-        MRI->getRegClass(Reg)->getID() == X86::TILERegClassID)
-      return true;
+    if (Reg.isVirtual()) {
```
[the remark this answers is not in the document]
Comment (phoebewang) [reply llvm/lib/Target/X86/X86FastTileConfig.cpp:97]: The function is in a different file. I don't find a proper header to store it and we should not call it across files.

File: llvm/lib/Target/X86/X86LowerAMXType.cpp [hunk 4]
```diff
@@ -919,23 +1017,66 @@ bool X86LowerAMXCast::optimizeAMXCastFromPhi(
...
 //                                           i64 64, x86_amx %42)
 bool X86LowerAMXCast::combineCastStore(IntrinsicInst *Cast, StoreInst *ST) {
   Value *Tile = Cast->getOperand(0);
-  // TODO: If it is cast intrinsic or phi node, we can propagate the
-  // shape information through def-use chain.
-  if (!isAMXIntrinsic(Tile))
+
+  assert(Tile->getType()->isX86_AMXTy() && "Not Tile Operand!");
+
+  // TODO: Specially handle the mult-use case.
```
Comment (fzou1) [inline llvm/lib/Target/X86/X86LowerAMXType.cpp:1043]: mult->multi. The same below.

File: clang/lib/Headers/amxtransposeintrin.h [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+  __builtin_ia32_t2rpntlvwz1(tdst, base, stride)
+#define _tile_2rpntlvwz1t1(tdst, base, stride)                                 \
+  __builtin_ia32_t2rpntlvwz1t1(tdst, base, stride)
+
+/// Transpose 32-bit elements from \a src and write the result to \a dst.
+///
+/// \headerfile <immintrin.h>
+///
+/// \code
+/// void __tile_transposed(__tile dst, __tile src);
```
[the remark this answers is not in the document]
Comment (phoebewang) [reply clang/lib/Headers/amxtransposeintrin.h]: Done.

File: clang/lib/Headers/amxtransposeintrin.h [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+/// \endcode
+///
+/// This intrinsic corresponds to the <c> TTRANSPOSED </c> instruction.
+///
+/// \param dst
+/// 	The destination tile. Max size is 1024 Bytes.
+/// \param src
+/// 	The 1st source tile. Max size is 1024 Bytes.
```
[the remark this answers is not in the document]
Comment (phoebewang) [reply clang/lib/Headers/amxtransposeintrin.h]: Done.

File: clang/lib/Headers/amxtransposeintrin.h [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+/// Transpose 32-bit elements from src and write the result to dst.
+///
+/// \headerfile <immintrin.h>
+///
+/// This intrinsic corresponds to the <c> TTRANSPOSED </c> instruction.
+///
+/// \param dst
+///    The destination tile. Max size is 1024 Bytes.
+/// \param src
+///    The 1st source tile. Max size is 1024 Bytes.
```
[the remark this answers is not in the document]
Comment (phoebewang) [reply clang/lib/Headers/amxtransposeintrin.h]: Done.

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 8]
```diff
@@ -16920,6 +16920,58 @@ Value *CodeGenFunction::EmitX86BuiltinExpr(unsigned BuiltinID,
...
+    Builder.CreateDefaultAlignedStore(VecT0, Ops[3]);
+
+    // Then store tile1 into DstPtr1
+    Value *T1 = Builder.CreateExtractValue(Call, 1);
+    Value *VecT1 = Builder.CreateIntrinsic(Intrinsic::x86_cast_tile_to_vector,
+                                           {TyPtee}, {T1});
+    Value *Store = Builder.CreateDefaultAlignedStore(VecT1, Ops[4]);
+
+    // Note: Here we escape directly use x86_tilestored64_internal to store
+    // the results due to it can't make sure the Mem writen scope. This may
```
[the remark this answers is not in the document]
Comment (phoebewang) [reply clang/lib/CodeGen/CGBuiltin.cpp]: Done.

File: llvm/include/llvm/CodeGen/TileShapeInfo.h [hunk 9]
```diff
@@ -34,9 +34,31 @@ class ShapeT {
...
+    for (auto *Shape : ShapesOperands)
+      Shapes.push_back(Shape);
+
+    if (MRI)
+      deduceImm(MRI);
+  }
   ShapeT()
       : Row(nullptr), Col(nullptr), RowImm(InvalidImmShape),
         ColImm(InvalidImmShape) {}
+  // TODO: We need to extern cmp operator for muti-shapes if
```
[the remark this answers is not in the document]
Comment (phoebewang) [reply llvm/include/llvm/CodeGen/TileShapeInfo.h]: Done.

File: llvm/include/llvm/CodeGen/TileShapeInfo.h [hunk 10]
```diff
@@ -34,9 +34,31 @@ class ShapeT {
     if (MRI)
       deduceImm(MRI);
   }
+  // When ShapeT has mult shapes, we only use Shapes (never use Row and Col)
```
[the remark this answers is not in the document]
Comment (phoebewang) [reply llvm/include/llvm/CodeGen/TileShapeInfo.h]: Done.

File: llvm/lib/Target/X86/AsmParser/X86Operand.h [hunk 11]
```diff
@@ -623,6 +623,37 @@ struct X86Operand final : public MCParsedAsmOperand {
     Inst.addOperand(MCOperand::createReg(Reg));
   }
 
+  bool isTILEPair() const {
+    return Kind == Register &&
+           X86MCRegisterClasses[X86::TILERegClassID].contains(getReg());
```
[the remark this answers is not in the document]
Comment (phoebewang) [reply llvm/lib/Target/X86/AsmParser/X86Operand.h]: Good catch, done.

File: llvm/lib/Target/X86/X86LowerAMXType.cpp [hunk 12]
```diff
@@ -121,12 +137,96 @@ static Instruction *getFirstNonAllocaInTheEntryBlock(Function &F) {
...
+  if (isa<ConstantInt>(V))
+    RealCol =
+        Builder.getInt16((cast<ConstantInt>(V)->getSExtValue()) * Granularity);
+  else if (isa<Instruction>(V)) {
+    Builder.SetInsertPoint(cast<Instruction>(V));
+    RealCol = Builder.CreateNUWMul(V, Builder.getInt16(Granularity));
+    cast<Instruction>(RealCol)->moveAfter(cast<Instruction>(V));
+  } else {
+    // When it is not a const value and it is a function argument, we create
+    // Row at the entry bb.
```
[the remark this answers is not in the document]
Comment (phoebewang) [reply llvm/lib/Target/X86/X86LowerAMXType.cpp:208]: Row is correct.

File: llvm/lib/Target/X86/X86LowerAMXType.cpp (hunk 4 above)
Comment (phoebewang) [reply llvm/lib/Target/X86/X86LowerAMXType.cpp]: Done.

**Review by fzou1 — APPROVED** [review]
LGTM

**Comment by nikic** [comment]
Looks like this causes a significant compile-time regression, but only for ReleaseLTO-g: https://llvm-compile-time-tracker.com/compare.php?from=1e19f0f9d92b5e9c43d53893e387341835d3d96b&to=c72a751dabff4260dcc309e48008941d51b31d21&stat=instructions:u

I wouldn't really expect this change to only affect that configuration -- I guess that means that optimized debuginfo processing has become a lot slower? Any idea what is going on here?

**Comment by phoebewang** [comment]
> Looks like this causes a significant compile-time regression, but only for ReleaseLTO-g: https://llvm-compile-time-tracker.com/compare.php?from=1e19f0f9d92b5e9c43d53893e387341835d3d96b&to=c72a751dabff4260dcc309e48008941d51b31d21&stat=instructions:u
> 
> I wouldn't really expect this change to only affect that configuration -- I guess that means that optimized debuginfo processing has become a lot slower? Any idea what is going on here?

Thanks for the report, should be solved by https://github.com/llvm/llvm-project/pull/114642

**Pull request merged**